Add new resolver `oc.create` (#677)
omry/omegaconf: docs/source/custom_resolvers.rst

    >>> show(cfg.database.password3)
    type: NoneType, value: None

.. _oc.create:

oc.create
^^^^^^^^^

``oc.create`` may be used for dynamic generation of config nodes
(typically from Python ``dict`` / ``list`` objects or YAML strings, similar to :ref:`OmegaConf.create<creating>`).
The following example combines ``oc.create`` with ``oc.decode`` and ``oc.env`` to generate
a sub-config from an environment variable:

.. doctest::

    >>> cfg = OmegaConf.create(
    ...     {
    ...         "model": "${oc.create:${oc.decode:${oc.env:MODEL}}}",
    ...     }
    ... )
    >>> os.environ["MODEL"] = "{name: my_model, layer_size: [100, 200]}"
    >>> show(cfg.model.layer_size)
    type: ListConfig, value: [100, 200]

> 

Reviewer: This seems a bit much (using 3 resolvers to explain 1).
Maybe something like this?
```python
OmegaConf.register_new_resolver("identity", lambda x: x)
cfg = OmegaConf.create(
    {
        "plain_dict": "${identity:{a:10}}",
        "dict_config": "${oc.create:{a:10}}",
        "dict_config_from_env": "${oc.create:${oc.env:YAML_ENV}}",
    }
)

os.environ["YAML_ENV"] = "A: 10\nb: 20\nC: ${.A}" 
show(cfg.plain_dict)
show(cfg.dict_config)
show(cfg.dict_config_from_env)
# interpolations works because this is a DictConfig
print(cfg.dict_config_from_env.C)
```

Author: Done in 8107676 (with a few small changes)

oc.decode
^^^^^^^^^

With ``oc.decode``, strings can be converted into their corresponding data types using the OmegaConf grammar.
This grammar recognizes typical data types like ``bool``, ``int``, ``float``, ``dict`` and ``list``,
e.g. ``"true"``, ``"1"``, ``"1e-3"``, ``"{a: b}"``, ``"[a, b, c]"``.

Note that:

> - When providing as input to ``oc.decode`` a string that is meant to be decoded into another string, in general
>   the input string should be quoted (since only a subset of characters are allowed by the grammar in unquoted

Reviewer: This is a bit cumbersome, maybe:
```suggestion
- Input strings provided to `oc.decode` should be quoted because only a subset of the characters is allowed in unquoted strings
```

(make sure it formats properly).

Reviewer: I also think it would be good to show decoding of a literal input string as well (not the result of another resolver).

Author: > This is a bit cumbersome, maybe:
> (...)

Done in 8ea1549 (with minor changes)

> I also think it would be good to show decoding of a literal input string as well (not the result of another resolver).

I'm a bit reluctant to do so because:
* It would break a bit the "flow" of the current example
* I can't think of a situation where one would want to use `oc.decode` on a literal input string

Do you have a specific suggestion in mind?

Reviewer: No, just a suggestion. we can leave it like that.
